test.py

```diff
@@ -842,26 +842,37 @@ async def run(self, options: argparse.Namespace) -> Test:
 
         loggerPrefix = self.mode + '/' + self.uname
         logger = LogPrefixAdapter(logging.getLogger(loggerPrefix), {'prefix': loggerPrefix})
-        async with self.suite.clusters.instance(logger) as cluster:
-            try:
-                cluster.before_test(self.uname)
-                logger.info("Leasing Scylla cluster %s for test %s", cluster, self.uname)
-                self.args.insert(0, "--host={}".format(cluster.endpoint()))
-                self.is_before_test_ok = True
-                cluster.take_log_savepoint()
-                status = await run_test(self, options)
-                cluster.after_test(self.uname)
-                self.is_after_test_ok = True
-                self.success = status
-            except Exception as e:
-                self.server_log = cluster.read_server_log()
-                self.server_log_filename = cluster.server_log_filename()
-                if self.is_before_test_ok is False:
-                    print("Test {} pre-check failed: {}".format(self.name, str(e)))
-                    print("Server log of the first server:\n{}".format(self.server_log))
-                    # Don't try to continue if the cluster is broken
-                    raise
-            logger.info("Test %s %s", self.uname, "succeeded" if self.success else "failed ")
+        try:
+            async with self.suite.clusters.instance(logger) as cluster:
+                try:
+                    cluster.before_test(self.uname)
+                    logger.info("Leasing Scylla cluster %s for test %s", cluster, self.uname)
+                    self.args.insert(0, "--host={}".format(cluster.endpoint()))
+                    self.is_before_test_ok = True
+                    cluster.take_log_savepoint()
+                    status = await run_test(self, options)
+                    cluster.after_test(self.uname)
+                    self.is_after_test_ok = True
+                    self.success = status
+                except Exception as e:
+                    self.server_log = cluster.read_server_log()
+                    self.server_log_filename = cluster.server_log_filename()
+                    if self.is_before_test_ok is False:
+                        print("Test {} pre-check failed: {}".format(self.name, str(e)))
+                        print("Server log of the first server:\n{}".format(self.server_log))
+                        # Don't try to continue if the cluster is broken
+                        raise
+                    if self.is_after_test_ok is False:
+                        print("Test {} pre-check failed: {}".format(self.name, str(e)))
+                        print("Server log of the first server:\n{}".format(self.server_log))
+                        logger.info(f"Discarding cluster after failed test %s...", self.name)
+                        await self.suite.clusters.steal()
```

Comment:
The fact that you are calling `steal()` on `clusters`, and not on the context manager, breaks the RAII abstraction suggested by the context manager. The `steal()` call should be done by the context manager.

Besides the abstraction, requiring both the `steal()` and the `raise` is error-prone — it's easy to forget the `steal`. They should happen together.

Reply:
Maybe a regular `try...except`, not RAII, could be the better abstraction here.